### PR TITLE
Add ES6 collection support to include()

### LINKED ISSUE #970: expect(...).to.include should support ES2015 Sets and Maps
It seems like Chai supports key assertions on ES2015 sets & maps (introduced in this PR: https://github.com/chaijs/chai/pull/633), but it does not support simple "inclusion" assertions.

Consider the following testcase:
```javascript
it("should detect item existence in Set", function () {
    let container = new Set();

    container.add(123);

    expect(container).to.include(123);
});
```

This test should obviously pass, as the ``container`` has ``123`` value added to it. Instead, it fails with the following error:
```
AssertionError: object tested must be an array, an object, or a string, but set given
```

Same thing happens if we change ``Set`` to ``Map`` (``container.set("somekey", 123)``).

---

As far as I know, this problem occurs both in ``3.5.0`` version, as well as the latest ``4.0.0-canary.2`` pre-release version.

### PR DESCRIPTION
**Pull request by shvaikalesh**
Closes #970

### CODE REVIEW

File: lib/chai/core/assertions.js [hunk 1]
```diff
@@ -504,10 +499,42 @@ module.exports = function (chai, _) {
...
+            undefined,
+            ssfi
+          );
+        }
+
+        included = obj.has(val);
+        break;
+
+      default:
+        var isEql = isDeep ? _.eql : function(a, b) { return a === b; };
```
Comment (shvaikalesh) [inline lib/chai/core/assertions.js]: I feel like we should use `SameValue` here, but it is minor breaking change.

File: lib/chai/core/assertions.js [hunk 2]
```diff
@@ -504,10 +499,42 @@ module.exports = function (chai, _) {
...
+      , included = false;
+
+    switch (objType) {
+      case 'string':
+        included = obj.indexOf(val) !== -1;
+        break;
+
+      case 'weakmap':
+      case 'weakset':
+        if (isDeep) {
```
Comment (shvaikalesh) [inline lib/chai/core/assertions.js]: I think we should consider containing code like this into a private method: it is easy to forget to put `flagMsg`, pass `ssfi`, or skip second argument.

Comment (keithamus) [inline lib/chai/core/assertions.js]: Agreed, but I'd like to see focus on #585 to get the assertion functions to be pure functions returning objects, so we can wrap them up in assertionerrors

File: lib/chai/core/assertions.js (hunk 1 above)
Comment (keithamus) [inline lib/chai/core/assertions.js]: Well this is a new feature right? So no one would have hit this path before. It would be a small gotcha to say:

> If you pass in (weak)?(map|set) then it'll use samevalue, otherwise ===

Comment (keithamus) [inline lib/chai/core/assertions.js]: Yes - we should look to change many of these instances for chai@5 i think

**Comment by keithamus** [comment]
I'm very happy with this, but let's get another review to sign off before merge. @meeber @lucasfcosta?

File: lib/chai/core/assertions.js [hunk 3]
```diff
@@ -504,10 +499,42 @@ module.exports = function (chai, _) {
...
+        }
+
+        included = obj.has(val);
+        break;
+
+      default:
+        var isEql = isDeep ? _.eql : function(a, b) { return a === b; };
+        obj.forEach(function (item) {
+          included = included || isEql(item, val);
+        });
```
Comment (vieiralucas) [inline lib/chai/core/assertions.js]: Can we do something here to optimize this by breaking out the loop once `included` is true?

I realize that then we would need to have 3 implementations, one for `Array`, another for `Map` and another for `Set`.

Do you think it is worth it?

Comment (shvaikalesh) [inline lib/chai/core/assertions.js]: We can optimize away looping through `map` and `set` for non-deep `include` (cool) and use `some` + `indexOf` on arrays. Code:

```diff
+  case 'map':
+  case 'set':
+    if (isDeep) {
+      obj.forEach(function (item) {
+        included = included || _.eql(item, val);
+      });
+   } else {
+     included = obj.has(val);
+   }
+   break;
+
+  case 'array':
+    if (isDeep) {
+     included = obj.some(function (item) {
+       return _.eql(item, val);
+     })
+    } else {
+      included = obj.indexOf(val) !== -1;
+    }
+    break;

-  default:
-    var isEql = isDeep ? _.eql : function(a, b) { return a === b; };
-    obj.forEach(function (item) {
-      included = included || isEql(item, val);
-    });
```

**Comment by shvaikalesh** [comment]
One more thing: should we throw when `deep.include` is called on string? Currently we just ignore the flag, but the assertion does not make sense.

File: lib/chai/core/assertions.js (hunk 3 above)
Comment (keithamus) [inline lib/chai/core/assertions.js]: This looks good but I'd be a little careful here with IE11 - which has terrible support for Map/Set. It probably will work but I've often been surprised how seemingly obvious features are missing from IE11's Map/Set

**Comment by keithamus** [comment]
@shvaikalesh I think we should - as a general future feature consideration - start adding warnings for wonky assertion chains, rather than explicitly erroring. Let's leave it out of this PR and discuss it in https://github.com/chaijs/chai/issues/996

**Comment by shvaikalesh** [comment]
I've updated the docs and dropped `WeakMap` support -- we can't iterate over it's values to make an assertion.

Regarding `deep` flag: point raised by @meeber in https://github.com/chaijs/chai/issues/996#issuecomment-310481585 is valid for `WeakSet` targets too. I am fine with silently ignoring it.

File: lib/chai/core/assertions.js [hunk 4]
```diff
@@ -504,10 +513,62 @@ module.exports = function (chai, _) {
...
+            ssfi
+          );
+        }
+
+        included = obj.has(val);
+        break;
+
+      case 'map':
+        var isEql = isDeep ? _.eql : SameValueZero;
+        obj.forEach(function (item) {
```
Comment (lucasfcosta) [inline lib/chai/core/assertions.js]: As @vieiralucas said, we might be able to optimize this by getting out of the loop when we found the included item.

In order to support IE and older versions of it we can `throw` an error inside the `forEach` callback and catch it.

This is a cross-platform compatible implementation I've made on `Sinon` for the `every` method. We can inspire our `some` implementation in that. However, even though I'd like it to be done, I won't block merging this PR because of that.

File: lib/chai/core/assertions.js [hunk 5]
```diff
@@ -330,6 +330,16 @@ module.exports = function (chai, _) {
    *
    *     expect({a: 1, b: 2, c: 3}).to.include({a: 1, b: 2});
    *
+   * When the target is a Set or WeakSet, `.include` asserts that the given `val` is a
+   * member of the target. SameValueZero equality algorithm is used.
+   *
+   *     expect(new Set([1, 2])).to.include(2);
+   *
+   * When the target is a Map, `.include` asserts that the given `val` is one of
+   * the values of the target. SameValueZero equality algorithm is used.
```
Comment (lucasfcosta) [inline lib/chai/core/assertions.js]: I think it would be good to explain how the `SameValueZero` algorithm works. I think many people could get confused by that and would end up either having to read the code to understand it or just avoiding using this assertion because they're not sure of what it does.

Comment (meeber) [inline lib/chai/core/assertions.js]: @lucasfcosta Fair point. In this case, I would suggest moving the phrase "SameValueZero equality algorithm is used" from where it is currently down to the separate paragraph where `strict` and `deep` equality` is discussed. This would make the doc style more consistent anyway. And then rather than explain what "SameValueZero equality" is, we could provide a link to an explanation, just like we do with the "deep equal" algorithm.

File: lib/chai/core/assertions.js (hunk 4 above)
Comment (meeber) [inline lib/chai/core/assertions.js]: @lucasfcosta I think this is worth exploring. I'd prefer this kind of optimization technique be implemented in a separate PR as a refactor.

File: lib/chai/core/assertions.js (hunk 5 above)
Comment (keithamus) [inline lib/chai/core/assertions.js]: We could just refer to is as it is more commonly known; [`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) or if you'd like to we coud keep the phrasing but link to [the equality article on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness) which mentions `SameValueZero`.

File: lib/chai/core/assertions.js (hunk 4 above)
Comment (lucasfcosta) [inline lib/chai/core/assertions.js]: @meeber agreed.

LGTM! Let's get this merged.
